specs/DesignDiscussions/Template.md

```diff
@@ -0,0 +1,418 @@
+
+## Parsing Strategy
+Current grammar is based on standard json grammar with addtional rules to handle adaptive card template language
+```json
+{
+	"$schema": "http://adaptivecards.io/schemas/adaptive-card.json",
+	"type": "AdaptiveCard",
+	"version": "1.0",
+	"body": [
+		{
+			"type": "TextBlock",
+			"text": "Lorem ipsum dolor sit amet",
+			"wrap": true
+		}
+	]
+}
+```
+The above json will be parsed into a parse tree by the parser.  
+ANTLR also provides a walker that takes a visitor as an argument. As the walker walks the tree, visitor's visit methods are called. Visit methods are defined per new grammar by default.
+When a visitor method is called, it will be called with a context. The context will be a data structure representing the current visiting node.
+
+```
+pair
+   : StringDeclOpen STRING CLOSE COLON value    # jsonPair
+   | StringDeclOpen TEMPLKEYWRD CLOSE COLON value # templateData
+   ;
+```
+
+From this grammar snippet, the parser generation tool will create a pairVisit method to a base visitor that we have to implement. By overriding this method, when a walker visits the pair node, it calls the visit method of the node. We can do our own work such as translation in the visit method.
+
+![](assets/Template/antlr4_parse_tree_1.png)
+**Figure 2**
+
+**Parse Tree with Templated String**
+```json
+{
+	"$schema": "http://adaptivecards.io/schemas/adaptive-card.json",
+	"type": "AdaptiveCard",
+	"version": "1.0",
+	"body": [
+		{
+			"type": "TextBlock",
+			"text": "{content}",
+			"wrap": true
+		}
+	]
+}
+```
+
+Notice that templateString node has been added, likewise other template keywords such as $data, $when, $root, $index can be parsed by the generated parser and lexer  
+Parser and lexer can be used on other platforms since they are generated by ANTLR tool and can be targeted to different languages. Visitor and Return Objects are needed to be implemented on other platforms.  
+  
+![](assets/Template/antlr4_parse_tree_2.png)
+**Figure 3**
+
+## Adaptive Expression Language
+This was initially intended to provide us functions support. AEL already has many built-in functions such as toLower, toString, equals, add, div, and [etc](https://github.com/microsoft/BotBuilder-Samples/blob/master/experimental/common-expression-language/prebuilt-functions.md)  
+But the AEL can do more. It can evaluate our templated string as well.
+More upto date information on AEL, please check [Matt's documentation](https://dev.azure.com/microsoft/AdaptiveCards/_git/Explorations?path=%2Fdocs%2FExplorations%2FTemplateLanguage%2FCommonExpressionLanguage.md&_a=preview) and [AEL's github repo](https://github.com/microsoft/BotBuilder-Samples/tree/master/experimental/common-expression-language).
+
+```dotnet
```

Comment:
You may want to use `csharp` here, since .NET is the framework, not the language